### PR TITLE
Add steps for volume discount program

### LINKED ISSUE #9445: [Feature]:  Add steps for volume discount program
# Feature Overview

**In order to** test the volume discount program in feature tests 
**We will** require new steps
**So that** we can propose programs, and check discount factors.

#### Step: "the volume discount tiers 'name':"
A step which can define volume discount tiers to be passed in a volume discount program governance proposal.
- mandatory columns
  - min running volume
  - discount factor

- [x] Done
 
#### Step: “the volume discount program:”
A step for proposing a volume discount program.
- mandatory columns
  - discount tiers (should be the name of either default or previously defined discount tiers)
  - closing timestamp
  - window length
- optional columns
  - err

- [x] Done

#### Step: “the volume discount stats at epoch "N":”
A step for checking a parties volume discount program stats at a specific epoch.
(Can be flexible in this step, I have just specced it in a fashion similar to the equivalent referral program step).
- mandatory fields
  - epoch
- mandatory columns
  - party
  - running volume
  - discount factor

- [x] Done

#### Step: "the following trades should be executed:"
Needs to be enhanced to optionally report applied volume discounts (this is already done but including for completeness).

- [x] Done

#### Step: "the following network parameters are set:"
Needs to be enhanced to cover the following:
- `volumeDiscountProgram.maxBenefitTiers`
- `volumeDiscountProgram.maxVolumeDiscountFactor`

- [x] Done

## Tasks
A checklist of the tasks that are needed to develop the feature and meet the acceptance criteria and feature test scenarios.
- [ ] implement cucumber steps for the volume discount program

### PR DESCRIPTION
**Pull request by ValentinTrinque**
Close #9445

### CODE REVIEW

File: datanode/entities/volume_discount.go [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+	"time"
+
+	"code.vegaprotocol.io/vega/libs/num"
+	eventspb "code.vegaprotocol.io/vega/protos/vega/events/v1"
+)
+
+type (
+	VolumeDiscountStatsUpdated struct {
+		AtEpoch                  uint64
+		PartyVolumeDiscountStats []*PartyVolumeDiscountStats
```
Comment (guoguojin) [inline datanode/entities/volume_discount.go:14]: If you don't need to, I think you can just use the proto message here instead of having to map it out.

Comment (ValentinTrinque) [reply datanode/entities/volume_discount.go:14]: happy to use it !

File: protos/sources/vega/events/v1/events.proto [hunk 2]
```diff
@@ -13,6 +13,24 @@ import "vega/vega.proto";
 
 option go_package = "code.vegaprotocol.io/vega/protos/vega/events/v1";
 
+// Stats of all parties eligible to volume discount.
```
Comment (candida-d) [inline protos/sources/vega/events/v1/events.proto:16]:
```suggestion
// Stats of all parties eligible for volume discount.
```

File: protos/sources/vega/events/v1/events.proto [hunk 3]
```diff
@@ -13,6 +13,24 @@ import "vega/vega.proto";
...
+message VolumeDiscountStatsUpdated {
+  // Epoch at which the volume discount statistics are updated.
+  uint64 at_epoch = 1;
+  // All parties stats.
```
Comment (candida-d) [inline protos/sources/vega/events/v1/events.proto:20]:
```suggestion
  // All parties' stats.
```

File: protos/sources/vega/events/v1/events.proto [hunk 4]
```diff
@@ -13,6 +13,24 @@ import "vega/vega.proto";
...
+  repeated PartyVolumeDiscountStats stats = 2;
+}
+
+// The volume discount stats for a given party.
+message PartyVolumeDiscountStats {
+  // The party.
```
Comment (candida-d) [inline protos/sources/vega/events/v1/events.proto:26]:
```suggestion
// Volume discount stats for a given party.
message PartyVolumeDiscountStats {
  // Party ID.
```

File: protos/sources/vega/events/v1/events.proto [hunk 5]
```diff
@@ -13,6 +13,24 @@ import "vega/vega.proto";
...
+  string party_id = 1;
+  // Discount factor applied to the volume.
+  string discount_factor = 2;
+  // Running volume of the party.
```
Comment (candida-d) [inline protos/sources/vega/events/v1/events.proto:30]:
```suggestion
  // The party's running volume.
```

**Review by guoguojin — DISMISSED** [review]

**Review by jeremyletang — DISMISSED** [review]

File: core/integration/steps/the_volume_discount_stats_should_be.go [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+			PartyId:        partyID,
+			DiscountFactor: specificRow.DiscountFactor(),
+			RunningVolume:  specificRow.RunningVolume(),
+		}
+	}
+
+	return stats, nil
+}
+
+func compareVolumeDiscountStats(expectedVolumeDiscountStats map[string]eventspb.PartyVolumeDiscountStats, foundVolumeDiscountStats map[string]eventspb.PartyVolumeDiscountStats) error {
```
Comment (EVODelavega) [inline core/integration/steps/the_volume_discount_stats_should_be.go:63]: nit-pick:
```go
func compareVolumeDiscountStats(expectedVolumeDiscountStats, foundVolumeDiscountStats map[string]eventspb.PartyVolumeDiscountStats) error {
```

So as to not repeat the rather long type name

File: core/integration/stubs/broker_stub.go [hunk 7]
```diff
@@ -936,6 +936,21 @@ func (b *BrokerStub) VestingStats() []eventspb.VestingStatsUpdated {
...
 }
 
+func (b *BrokerStub) VolumeDiscountStats() []eventspb.VolumeDiscountStatsUpdated {
+	batch := b.GetBatch(events.VolumeDiscountStatsUpdatedEvent)
+
+	stats := make([]eventspb.VolumeDiscountStatsUpdated, 0, len(batch))
+	for _, event := range batch {
+		switch et := event.(type) {
+		case *events.VolumeDiscountStatsUpdated:
+			stats = append(stats, et.Proto())
```
Comment (EVODelavega) [inline core/integration/stubs/broker_stub.go:946]: Probably out of scope of this PR, but the `events.Proto()` functions are inconsistent WRT their return values. `Party.Proto()` returns `*proto.Party`, whereas most other types like `RiskFactor.Proto()` return `proto.RiskFactor`. There may be a good reason for this behaviour, but if there isn't, we probably want to make this consistent at some point.

Comment (ValentinTrinque) [reply core/integration/stubs/broker_stub.go:946]: Created the issue.

File: core/integration/steps/the_volume_discount_stats_should_be.go [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+			PartyId:        partyID,
+			DiscountFactor: specificRow.DiscountFactor(),
+			RunningVolume:  specificRow.RunningVolume(),
+		}
+	}
+
+	return stats, nil
+}
+
+func compareVolumeDiscountStats(expectedStats map[string]eventspb.PartyVolumeDiscountStats, foundStats map[string]eventspb.PartyVolumeDiscountStats) error {
```
Comment (EVODelavega) [inline core/integration/steps/the_volume_discount_stats_should_be.go:63]: repeated arg types

**Review by jeremyletang — APPROVED** [review]

**Pull request merged**